shaka/src/media/pipeline_monitor.cc

```diff
@@ -69,14 +69,56 @@ void PipelineMonitor::Stop() {
   thread_.join();
 }
 
+bool AtEnd(const BufferedRanges& ranges, double time, double duration) {
+  if (std::isnan(duration)) {
+    return false;
+  }
+
+  const double relative_time = time - duration;
+
+  if (relative_time >= 0) {
+    return true;
+  }
+
+  // With presentation time tracked with the help of audio device/data a
+  // possible scenario is that audio track is effectively shorter than
+  // advertised duration, or implicit duration defined by video track. In such
+  // circumstances playback time is unable to reach duration exactly. Being
+  // |kNearEndRangeThreshold| close to end of playback, we assume end of
+  // presentation on reaching end time of last buffered range (which itself is
+  // an intersection of ranges of tracks). Threshold of |kAssumeEndThreshold|
+  // compensates for possible discrepancy of external clock processing last
+  // packet with respective update of presentation time.
+  //
+  // To be even more precise and ensure exactly last video frame is displayed
+  // from available data, we could possibly notify external clock of expected
+  // duration and allow media time reporting past its data derived from audio
+  // track. In this case audio playback subsystem having no real data but with
+  // known duration would be able to use audio hardware defined clock to
+  // continue advancing time up to duration value and take playback exactly to
+  // duration time.
+  static constexpr const double kNearEndRangeThresholdS = 0.500;
+  static constexpr const double kAssumeEndThresholdS = 0.025;
```

Comment:
Since this is related to the buffered ranges, I would suggest using `FrameBuffer::kMaxGapSize`.  That is used for fudging the end elsewhere too.